### DISCUSSION TITLE
How to customize the +prefix menus in Keypad mode?

### DISCUSSION THREAD

**Discussion by pauljamesharper** [meow-edit/meow#649] [General]
How do you customize the +prefix menu's in Keypad mode? For example in the photo below the "t" +prefix should be "themes" and the "d" +prefix should be "denote".

![Screenshot from 2024-10-20 18-01-43](https://github.com/user-attachments/assets/a7e04b85-51a3-49de-a481-e21f6577f3e2)

**Comment by DogLooksGood** [comment]
You should use an alias for the keymap, check my answer here:

https://github.com/meow-edit/meow/issues/627

**Comment by pauljamesharper** [comment]
Thanks for that reply.  It gave me information to go digging further. I found this link from spook.
[Helper function to create prefix keys for leader commands #131](https://github.com/meow-edit/meow/pull/131) I have denote and modus theme changing working. (Although they still appear after 'w' +prefix still follow by 'd' +prefix and 't' +prefix respectively.) My config looks like this. 

`**  Meow customizations
From: https://github.com/meow-edit/meow/pull/131

Macro
#+begin_src emacs-lisp
   (defmacro spook-defkeymap (name prefix &rest bindings)
    "Create a new NAME-keymap bound to PREFIX, with BINDINGS.

  Usage:
    (spook-defkeymap \"spook-git\" \"C-c g\"
      '(\"s\" . magit-status))
  "
    (let* ((keymap-name (intern (concat name "-keymap")))
           (keymap-alias (intern name))
           (keymap-bindings (mapcar
                             (lambda (binding)
                               (let ((binding (eval binding)))
                                 `(define-key ,keymap-name (kbd ,(car binding)) #',(cdr binding))))
                             bindings)))
      `(progn
         (defvar ,keymap-name (make-sparse-keymap))
         (defalias ',keymap-alias ,keymap-name)
         (global-set-key (kbd ,prefix) ',keymap-alias)
         ,@keymap-bindings)))
#+end_src
I define keymap like this:

#+begin_src emacs-lisp
    (spook-defkeymap
     "themes" "C-c t"
     '("t" . modus-themes-toggle)
     '("m" . modus-themes-select)
     '("s" . consult-theme))
    (spook-defkeymap
     "notes" "C-c d"
     '("b" . denote-find-backlink)
     '("d" . denote-date)
     '("f" . denote-find-link)
     '("h" . denote-org-extras-link-to-heading)
     '("i" . denote-link-or-create)
     '("k" . denote-rename-file-keywords)
     '("l" . denote-insert-link)
     '("n" . denote)
     '("p" . ews-denote-para-keywords)
     '("r" . denote-rename-file)
     '("R" . denote-rename-file-using-front-matter)
     '("C" . consult-notes)
     '("g" . consult-notes-search-in-all-notes))

#+end_src

Then I assign them to a meow-leader-key:

#+begin_src emacs-lisp
    (meow-leader-define-key
     '("t" . themes)
     '("d"**  Meow customizations
From: https://github.com/meow-edit/meow/pull/131

Macro
#+begin_src emacs-lisp
   (defmacro spook-defkeymap (name prefix &rest bindings)
    "Create a new NAME-keymap bound to PREFIX, with BINDINGS.

  Usage:
    (spook-defkeymap \"spook-git\" \"C-c g\"
      '(\"s\" . magit-status))
  "
    (let* ((keymap-name (intern (concat name "-keymap")))
           (keymap-alias (intern name))
           (keymap-bindings (mapcar
                             (lambda (binding)
                               (let ((binding (eval binding)))
                                 `(define-key ,keymap-name (kbd ,(car binding)) #',(cdr binding))))
                             bindings)))
      `(progn
         (defvar ,keymap-name (make-sparse-keymap))
         (defalias ',keymap-alias ,keymap-name)
         (global-set-key (kbd ,prefix) ',keymap-alias)
         ,@keymap-bindings)))
#+end_src
I define keymap like this:

#+begin_src emacs-lisp
    (spook-defkeymap
     "themes" "C-c t"
     '("t" . modus-themes-toggle)
     '("m" . modus-themes-select)
     '("s" . consult-theme))
    (spook-defkeymap
     "notes" "C-c d"
     '("b" . denote-find-backlink)
     '("d" . denote-date)
     '("f" . denote-find-link)
     '("h" . denote-org-extras-link-to-heading)
     '("i" . denote-link-or-create)
     '("k" . denote-rename-file-keywords)
     '("l" . denote-insert-link)
     '("n" . denote)
     '("p" . ews-denote-para-keywords)
     '("r" . denote-rename-file)
     '("R" . denote-rename-file-using-front-matter)
     '("C" . consult-notes)
     '("g" . consult-notes-search-in-all-notes))

#+end_src

Then I assign them to a meow-leader-key:

#+begin_src emacs-lisp
    (meow-leader-define-key
     '("t" . themes)
     '("d" . notes

#+end_src`

**Comment by DogLooksGood** [comment]
Yeah, using alias for keymap is not a hacky thing, you can find the usage in emacs source code. 
https://github.com/emacs-mirror/emacs/blob/master/lisp/kmacro.el#L210

And, even though there's a `meow-leader-define-key`, it's for backward compatibility. I prefer this way
```
(keymap-set global-map "C-c t" 'themes)
(keymap-set global-map "C-c d" 'notes)
```
It works with/without Meow.

**Reply by pauljamesharper** [reply]
Thanks! the above works as described. The only issue I have is what if I needed to nest +prefix?

ie. C-c -n  for notes, then under that C-c n d for denote, C-c n x for Denote-Explore, C-c n c for Consult-Denote, etc.

Is that doable?

**Reply by DogLooksGood** [reply]
Then you need do the same way to each nested keymap.

```
(defvar notes-keymap (make-keymap))
(defalias 'notes-keymap notes-keymap)   ; defalias
(keymap-set global-map "C-c n" 'notes-keymap) ; use alias

(defvar denote-keymap (make-keymap))
(defalias 'denote-keymap denote-keymap) ;defalias
(keymap-set notes-keymap "d" 'denote-keymap) ;use alias
```

**Comment by pauljamesharper** [comment]
Thanks for the assistance so far. I am getting fairly usable keybinds. However, with General Key you can more verbose and descriptive items for the menus like _"Open Dired"_  or "_Dired Jump to current"_ after **:wk.**  Is there a way I can do similar with 
**my-defkeymap**? Or is this something unique to General?

```
;; Example of General Key
   (start/leader-keys
    "d" '(:ignore t :wk "Dired")
    "d v" '(dired :wk "Open dired")
    "d j" '(dired-jump :wk "Dired jump to current"))

;; Example from my Emacs meow config
(my-defkeymap
    "dired-file-manager" "C-c d"
    '("d" . dired)
    '("j" . dired-jump )) 
```

**Reply by DogLooksGood** [reply]
I think it's unique to general. The idea of meow is to leverage the default keybindings. 

However, the `SPC` in meow is just a simple command which takes you to the keypad state.  If you already have a well-defined leader system, just replace the `SPC` with your keymap.

```
(keymap-set meow-normal-state-keymap "SPC" 'your-keymap)
(keymap-set meow-motion-state-keymap "SPC" 'your-keymap)
```

**Comment by pauljamesharper** [comment]
I think I have everything working the way I want. My idea was to get my config similar to the starter kit [kickstart.emacs](https://github.com/MiniApollo/kickstart.emacs/tree/main). That starter kit uses Vim and General. 

I have more or less the same items keymaps in my keymaps. But some of the kickstart.emacs bindings are unnecassary like  SPC . to find-file. SPC  x f does the same and uses built in commands. 

The bit about verbose description in the menus I eventually worked out. You can see the pattern below.  You probably can guess what I know about programming can be written on a postage stamp. 

![menusmeow](https://github.com/user-attachments/assets/696b00e5-2b45-4b14-a02b-a19ffcc24ede)

I forked the starter kit above and cleverly named it [kickstart.meow](https://github.com/pauljamesharper/kickstart.meow/tree/main). It might be handy for someone who quickly wants to try out meow and/or develop a set of keymap bindings. 


```
(my-defkeymap
    "find" "C-c f"
    '("c" . ("Edit emacs config" . (lambda () (interactive) (find-file "~/.config/emacs/config.org"))))
    '("r" . ("Recent Files" . consult-recent-file))
    '("f" . ("Fd search for files"consult-fd))
    '("g" . ("Ripgrep search in files" . consult-ripgrep))
    '("l" . ("Find Line" . consult-line))
    '("i" . ("Imenu buffer locations" . consult-imenu)))
```

**Reply by DogLooksGood** [reply]
It looks nice.  I'm still using the default key bindings from every package and I'm still thinking something about [this](https://github.com/meow-edit/meow/issues/64).  Do a key translation will allow keypad to work with any other key binding utilities.

**Comment by brabalan** [comment]
I found a very interesting approach (using vanilla emacs) in [this video](https://www.youtube.com/watch?v=gojOZ3k1mmk). The relevant config (updated since the video as bindings can take a description) is [here](https://protesilaos.com/emacs/dotemacs#h:f860bd85-671e-477e-8233-23a8ba43e1c6).
